src/qibolab/instruments/zhinst.py

```diff
@@ -0,0 +1,1050 @@
+                                measure_pulse_amplitude=None,
+                                acquire_delay=self.time_of_flight * 1e-9,
+                                reset_delay=relaxation_time * 1e-9,
+                            )
+                            i += 1
+
+        if optimized:
+            print("I'm using optimized IW")
+        else:
+            print("I'm using dumb IW")
+
+    def fast_reset(self, exp, qubits, fast_reset):
+        """
+        Conditional fast reset after readout - small delay for signal processing
+        This is a very naive approach that can be improved by repeating this step until
+        we reach non fast reset fidelity
+        https://quantum-computing.ibm.com/lab/docs/iql/manage/systems/reset/backend_reset
+        """
+        print("Im fast resetting")
+        for qubit_name in self.sequence_qibo.qubits:
+            qubit = qubits[qubit_name]
+            if not qubit.flux_coupler:
+                with exp.section(uid=f"fast_reset{qubit.name}", play_after=f"sequence_measure{qubit.name}"):
+                    with exp.match_local(handle=f"sequence{qubit.name}"):
+                        with exp.case(state=0):
+                            pass
+                        with exp.case(state=1):
+                            exp.play(signal=f"drive{qubit.name}", pulse=ZhPulse(fast_reset[qubit.name]).zhpulse)
+
+    def sweep(self, qubits, sequence, options, *sweepers):
+        """Play pulse and sweepers sequence"""
+
+        self.signal_map = {}
+
+        sweepers = list(sweepers)
+
+        dimensions = []
+        if options.averaging_mode is AveragingMode.SINGLESHOT:
+            dimensions = [options.nshots]
+
+        for sweeper in sweepers:
+            dimensions.append(len(sweeper.values))
+
+        # Re-arranging sweepers based on hardware limitations
+        # FIXME: Punchout and frequency case
+        rearranging_axes = [[], []]
+        if len(sweepers) == 2:
+            if sweepers[1].parameter is Parameter.frequency:
+                if sweepers[0].parameter is Parameter.bias:
+                    rearranging_axes[0] += [sweepers.index(sweepers[1])]
+                    rearranging_axes[1] += [0]
+                    sweeper_changed = sweepers[1]
+                    sweepers.remove(sweeper_changed)
+                    sweepers.insert(0, sweeper_changed)
+                    warnings.warn("Sweepers were reordered")
+                elif (
+                    not sweepers[0].parameter is Parameter.amplitude
+                    and sweepers[0].pulses.type is not PulseType.READOUT
+                ):
+                    rearranging_axes[0] += [sweepers.index(sweepers[1])]
+                    rearranging_axes[1] += [0]
+                    sweeper_changed = sweepers[1]
+                    sweepers.remove(sweeper_changed)
+                    sweepers.insert(0, sweeper_changed)
+                    warnings.warn("Sweepers were reordered")
```

Comment:
I had some trouble because if sweepers[0].parameter is Parameter.bias then sweepers[0].pulses will be None and it will not have .type and crash. Will it still run the or(...) if the first part is already true ?

Reply:
I will check it

Reply:
In Python, `or` and `and` are short-circuit operators: if `a is True` in `a or b`, then `b` is not evaluated. Same thing if `a is False` in `a and b`.

https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not

Reply:
great